### PR TITLE
Added toStrict to WS Messages (#15).

### LINKED ISSUE #15: Should ws.Message have toStrict() same as entities?
The same issue around "but I want it strict" surfaces in WebSocket messages as it does with HttpEntities - for entities people can easily toStrict them - for websocket messages where toStricting usually is more safe (smaller msgs... generally speaking), we don't have this helper on the type directly.

I think it would make a nice addition, opinions? Did I miss something?

### PR DESCRIPTION
**Pull request by rbudzko**
Please, take a look into concerns posted in #15. If fine - resolves #15.

### CODE REVIEW

**Comment by rbudzko** [comment]
PLS BUILD

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/ws/Message.scala [hunk 1]
```diff
@@ -68,9 +88,22 @@ sealed trait BinaryMessage extends akka.http.javadsl.model.ws.BinaryMessage with
...
 
+  /**
+   * Collects all possible parts and returns a potentially future Strict Message for easier processing.
+   * The Future is failed with an TimeoutException if the stream isn't completed after the given timeout.
+   */
+  def toStrict(timeout: FiniteDuration)(implicit fm: Materializer): Future[BinaryMessage.Strict] =
+    this match {
+      case BinaryMessage.Strict(binary) ⇒ Future.successful(BinaryMessage.Strict(binary))
+      case BinaryMessage.Streamed(binaryStream) ⇒ binaryStream
+        .runFold(ByteString())(_ ++ _)
```
Comment (viktorklang) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/ws/Message.scala]: Use ByteStringBuilder instead.

File: akka-http-core/src/test/scala/akka/http/impl/engine/ws/MessageSpec.scala [hunk 2]
```diff
@@ -849,6 +851,41 @@ class MessageSpec extends FreeSpec with Matchers with WithMaterializerSpec with
...
       }
     }
+    "convert any message to strict message" - {
+      val msg = "JKS*s';@"
+      val pre = "pre"
+      val post = "post"
+      "convert streamed text with multiple elements to strict text" in {
+        Await.result(
+          TextMessage
+            .Streamed(Source.fromIterator(() ⇒ Iterator(pre, msg, post)))
```
Comment (viktorklang) [inline akka-http-core/src/test/scala/akka/http/impl/engine/ws/MessageSpec.scala]: `Source(pre :: msg :: post :: Nil)`?

File: akka-http-core/src/test/scala/akka/http/impl/engine/ws/MessageSpec.scala [hunk 3]
```diff
@@ -849,6 +851,41 @@ class MessageSpec extends FreeSpec with Matchers with WithMaterializerSpec with
...
+            .toStrict(1.second), 1.second
+        ) should be(TextMessage.Strict(pre + msg + post))
+      }
+      "convert streamed binary with multiple elements to strict binary" in {
+        Await.result(
+          BinaryMessage
+            .Streamed(Source.fromIterator(() ⇒
+              Iterator(ByteString(pre.getBytes), ByteString(msg.getBytes), ByteString(post.getBytes))
```
Comment (viktorklang) [inline akka-http-core/src/test/scala/akka/http/impl/engine/ws/MessageSpec.scala]: Never use getBytes without charset

File: akka-http-core/src/test/scala/akka/http/impl/engine/ws/MessageSpec.scala [hunk 4]
```diff
@@ -849,6 +851,41 @@ class MessageSpec extends FreeSpec with Matchers with WithMaterializerSpec with
...
+        ) should be(TextMessage.Strict(msg))
+      }
+      "convert streamed binary to strict binary" in {
+        Await.result(
+          BinaryMessage
+            .Streamed(Source.single(ByteString(msg.getBytes)))
+            .toStrict(1.second), 1.second
+        ) should be(BinaryMessage.Strict(ByteString(msg.getBytes)))
+      }
+    }
```
Comment (viktorklang) [inline akka-http-core/src/test/scala/akka/http/impl/engine/ws/MessageSpec.scala]: Nothing tests the timeout.

There is no test for empty sequence.

**Comment by rbudzko** [comment]
Note to myself: As pointed out by @viktorklang - timeout to be implemented... Maybe with stage like `util.ToStrict` for `HttpEntity`.

**Comment by ktoso** [comment]
I'll review in depth tomorrow, thanks a lot for picking this one up @rbudzko ! :-)

**Comment by ktoso** [comment]
Not ready does not mean there's nothing to review though :-) 
Thanks for the heads up about your travel!

**Comment by ktoso** [comment]
PLS BUILD

**Comment by ktoso** [comment]
Mima filters:

```
[info] akka-http-core: found 6 potential binary incompatibilities while checking against com.typesafe.akka:akka-http-core_2.11:10.0.0  (filtered 18)
[error]  * abstract method toStrict(Long,akka.stream.Materializer)java.util.concurrent.CompletionStage in class akka.http.javadsl.model.ws.TextMessage is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.javadsl.model.ws.TextMessage.toStrict")
[error]  * abstract method toStrict(Long,akka.stream.Materializer)java.util.concurrent.CompletionStage in class akka.http.javadsl.model.ws.BinaryMessage is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.javadsl.model.ws.BinaryMessage.toStrict")
[error]  * method toStrict(scala.concurrent.duration.FiniteDuration,akka.stream.Materializer)scala.concurrent.Future in trait akka.http.scaladsl.model.ws.TextMessage is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.model.ws.TextMessage.toStrict")
[error]  * method toStrict(Long,akka.stream.Materializer)java.util.concurrent.CompletionStage in trait akka.http.scaladsl.model.ws.TextMessage is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.model.ws.TextMessage.toStrict")
[error]  * method toStrict(scala.concurrent.duration.FiniteDuration,akka.stream.Materializer)scala.concurrent.Future in trait akka.http.scaladsl.model.ws.BinaryMessage is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.model.ws.BinaryMessage.toStrict")
[error]  * method toStrict(Long,akka.stream.Materializer)java.util.concurrent.CompletionStage in trait akka.http.scaladsl.model.ws.BinaryMessage is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.model.ws.BinaryMessage.toStrict")
```

**Comment by ktoso** [comment]
Should be safe.

**Comment by jrudolph** [comment]
I think the change does what it advertises. However, I'd like to experiment with alternative APIs, instead, before people start to use `toStrict` (which requires to deal with futures) when an alternative could solve the issue in a simpler way. 

What I'd like to try first is adding a method to `UpgradeToWebSocket`:

```scala
def handleStrictMessages(Graph[FlowShape[StrictMessage, Message], Any], ...)
def handleStrictTextMessages(Graph[FlowShape[TextMessage.Strict, Message], Any], ...)
def handleStrictBinaryMessages(Graph[FlowShape[BinaryMessage.Strict, Message], Any], ...)
```

where `StrictMessage` is either a new hierarchy of WS messages or a new subtype. All of these should automatically "do the right thing", i.e. accumulating streamed messages with a configured timeout and maximum size and ignore messages of the wrong shape. We could probably reuse your `toStrict` logic for it. Is that something you would like to help with, @rbudzko?

**Comment by rbudzko** [comment]
@jrudolph , thank you for a feedback. I'll look into the matter when possible. As always, if the issue is burning, do not wait for my contribution.

**Comment by rbudzko** [comment]
It is unlikely we will need this PR.

**Comment by ktoso** [comment]
I really like what @jrudolph is proposing, but the toStrict would be good for two reasons:
- to have it now, until we actually do the other one
- for plain consistency, time and time again this catches me and others off guard that it's missing.

Can we reopen, add this, and then work on the new APIs please?

**Comment by ktoso** [comment]
cc @rbudzko @patriknw @raboof // FYI @jrudolph

**Comment by ktoso** [comment]
I'll revive this as separate PR since can't push into @rbudzko's branch; credits will remain proper ofc

**Pull request closed without merging**